lwm2m_client and utils: Location assistance object and usage (#7259)
nrfconnect/sdk-nrf: samples/nrf9160/lwm2m_client/location_assistance.rst

.. _location_assistance:

Location Assistance
===================

> The sample supports location assistance from nRF Cloud. This is achieved with using lwm2m-objects Signal Measurement Information (ID 10256) and Location assistance (ID 50001).

Reviewer: ```suggestion
The sample supports proprietary mechanism to fetch location assistance data from nRF Cloud by proxying it through LwM2M. This is achieved with using lwm2m-objects Signal Measurement Information (ID 10256) and proprietary Location assistance object (ID 50001).
```

> Currently the Coiote lwm2m-server can be used for utilizing the location assistance with nRF Cloud.

Reviewer: ```suggestion
This feature is currently under development and right now, only the AVSystem Coiote lwm2m-server can be used for utilizing the location assistance with nRF Cloud.
```

Reviewer 2: Made this a note.

There are currently two different methods of obtaining the location assistance.

First is the location based on cell information. The device sends the information about the current cell and possibly also neighbouring cells to the lwm2m-server.
The lwm2m-server then sends the location request to the nRF Cloud which responds with the location.

> Second method is requesting A-GPS assistance data, which is queried from the nRF Cloud and fed through the lwm2m-server back to the device. The A-GPS assistance data is then fed to the GNSS-module for obtaining a location faster.

Reviewer: ```suggestion
Second method is requesting A-GPS assistance data, which is queried from the nRF Cloud and fed through the lwm2m-server back to the device. The A-GPS assistance data is then fed to the GNSS-module for obtaining a position fix faster.
```

Reviewer 2: Updated

This is achieved by using LwM2M objects ECID-Signal Measurement Information object(ID 10256) and proprietary Location assistance object (ID 50001).

.. note::
   This feature is currently under development and as of now, only AVSystem's Coiote LwM2M server can be used for utilizing the location assistance data from nRF Cloud.

There are two different methods of obtaining the location assistance:

* Location based on cell information - The device sends information about the current cell and possibly about the neighboring cells to the  LwM2M server. The LwM2M server then sends the location request to nRF Cloud, which responds with the location data.
  Use :file:`overlay-assist-cell.conf` to enable this functionality in the sample.

* Query of A-GPS assistance data - The A-GPS assistance data is queried from nRF Cloud and provided back to the device through the LwM2M server. The A-GPS assistance data is then provided to the GNSS module for obtaining the position fix faster.
  Use :file:`overlay-assist-agps.conf` to enable this functionality in the sample.

Location assistance options
===========================

> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL`
>   Configuration for using cell based location as a location assistance method.
>   nRF Cloud estimates the location of the device based on the cell neighborhood.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS`
>   Configuration for using A-GPS as location assistance method
>   nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS`
>   Configuration for using events from :ref:`lib_lwm2m_client_utils` library for sending assistance requests
>   Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER`
>   Configuration for using a listener from :ref:`lib_lwm2m_client_utils` library for populating the cell neighborhood
>   Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.

Reviewer 2: Suggesting to remove the descriptions ( like 'Configuration for using cell based location as a location assistance method.') that are already present in the Kconfig description.
```suggestion
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` - nRF Cloud estimates the location of the device based on the cell neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS` -  nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER` - Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.
```

> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL`
>   Configuration for using cell based location as a location assistance method.
>   nRF Cloud estimates the location of the device based on the cell neighborhood.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS`
>   Configuration for using A-GPS as location assistance method
>   nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS`
>   Configuration for using events from :ref:`lib_lwm2m_client_utils` library for sending assistance requests
>   Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
> 
> * :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER`
>   Configuration for using a listener from :ref:`lib_lwm2m_client_utils` library for populating the cell neighborhood
>   Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.

Reviewer 2: Suggesting to remove the descriptions ( like 'Configuration for using cell based location as a location assistance method.') that are already present in the Kconfig description.
```suggestion
Check and configure the following library options that are used by the sample:

* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_CELL` - nRF Cloud estimates the location of the device based on the cell neighborhood.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_AGPS` -  nRF Cloud provides A-GPS assistance data and the GNSS-module in the device uses the data for obtaining a GNSS fix, which is reported back to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_LOCATION_ASSIST_EVENTS` - Disable this option if you provide your own method of sending the assistance requests to the LwM2M server.
* :kconfig:option:`CONFIG_LWM2M_CLIENT_UTILS_NEIGHBOUR_CELL_LISTENER` - Disable this option if you provide your own method of populating the LwM2M objects (ID 10256) containing the cell neighborhood information.
```
